mne/beamformer/_lcmv.py

```diff
@@ -39,33 +50,46 @@ def _reg_pinv(x, reg):
     return linalg.pinv(x), d
 
 
-def _setup_picks(picks, info, forward, noise_cov=None):
-    if picks is None:
-        picks = pick_types(info, meg=True, eeg=True, ref_meg=False,
-                           exclude='bads')
+def _setup_picks(info, forward, data_cov=None, noise_cov=None):
+    # get a list of all channel names:
+    fwd_ch_names = forward['info']['ch_names']
 
-    ok_ch_names = set([c['ch_name'] for c in forward['info']['chs']])
-    if noise_cov is not None:
-        ok_ch_names.union(set(noise_cov.ch_names))
+    # handle channels from forward model and info:
+    ch_names = _compare_ch_names(info['ch_names'], fwd_ch_names, info['bads'])
 
-    if noise_cov is not None and set(info['bads']) != set(noise_cov['bads']):
+    # inform about excluding channels:
+    if (data_cov is not None and set(info['bads']) != set(data_cov['bads']) and
+            (len(set(ch_names).intersection(data_cov['bads'])) > 0)):
+        logger.info('info["bads"] and data_cov["bads"] do not match, '
+                    'excluding bad channels from both.')
+    if (noise_cov is not None and
+            set(info['bads']) != set(noise_cov['bads']) and
+            (len(set(ch_names).intersection(noise_cov['bads'])) > 0)):
         logger.info('info["bads"] and noise_cov["bads"] do not match, '
                     'excluding bad channels from both.')
 
-    bads = set(info['bads'])
-    if noise_cov is not None:
-        bads.union(set(noise_cov['bads']))
-
-    ok_ch_names -= bads
+    # handle channels from data cov if data cov is not None
+    # Note: data cov is supposed to be None in tf_lcmv
+    if data_cov is not None:
+        ch_names = _compare_ch_names(ch_names, data_cov.ch_names,
+                                     data_cov['bads'])
 
-    ch_names = [info['chs'][k]['ch_name'] for k in picks]
-    ch_names = [c for c in ch_names if c in ok_ch_names]
+    # handle channels from noise cov if noise cov available:
+    if noise_cov is not None:
+        ch_names = _compare_ch_names(ch_names, noise_cov.ch_names,
+                                     noise_cov['bads'])
 
     picks = [info['ch_names'].index(k) for k in ch_names if k in
              info['ch_names']]
     return picks
 
 
+def _compare_ch_names(names1, names2, bads):
+    """Return channel names common to both lists but not in bads"""
+    ch_names = [ch for ch in names1 if ch not in bads and ch in names2]
+    return ch_names
+
+
 def _check_one_ch_type(info, picks, noise_cov):
```

Comment:
"""Add a mini 1 line docstring here that is a grammatically complete sentence.""""